browser/brave_shields/ad_block_subscription_download_manager_getter.cc

```diff
@@ -86,6 +86,10 @@ class AdBlockSubscriptionDownloadManagerGetterImpl
     }
   }
 
+  ~AdBlockSubscriptionDownloadManagerGetterImpl() override {
+    g_browser_process->profile_manager()->RemoveObserver(this);
```

Comment:
I'm not sure this will help, atm we either `RemoveObserver()` before `delete this;` or we simply don't do `AddObserver()`. And there is no other way to call the destructor other than `GetDownloadManager()` that does `delete this;` (sorta unusual side effect for a method named `Get...` btw).

Also with this we will remove an observer that wasn't added, if a profile is already in place.

Reply:
I think this is the first time see a destructor called from inside a constructor, see `AdBlockSubscriptionDownloadManagerGetterImpl` -> `GetDownloadManager` -> `delete this;`

Reply:
> with this we will remove an observer that wasn't added, if a profile is already in place.

@bridiver suggested this approach, supposedly it's fine to remove an observer that doesn't exist.

> a destructor called from inside a constructor

Interesting observation, but I don't think it's problematic? That only happens when the profile already exists, in which case there's no need for the instance to continue existing.

Reply:
> @bridiver suggested this approach, supposedly it's fine to remove an observer that doesn't exist.

it is fine, but as I've mentioned, this change doesn't fix the potential problem

Reply:
@iefremov I'm not sure I'm following. There is definitely an issue where the observer can get added, but may not get removed before this object is deleted. This changes ensures that the observer is always removed.

Reply:
I guess I do see a problem where `delete this` gets called from the constructor which is a separate issue and we can also fix that

Reply:
>  This changes ensures that the observer is always removed.

it does not, because (1) the observer is not removed if the destructor is not called (2) the destructor is not called if `OnProfileAdded()` is not called (3) there is a chance that `profile_manager()` is destroyed before calling `OnProfileAdded()`.

I'm trying to explain that the current change essentially does nothing, because at the moment we **already always call `RemoveObserver()`** before destroying the observer. With this change we will call `RemoveObserver()` **twice** if it was added and once if it wasn't.

Reply:
`GetDownloadManager` is now posted as a task, so that should be fixed